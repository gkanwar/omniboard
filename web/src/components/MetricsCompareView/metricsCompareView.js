import React, {Component} from 'react';
import PropTypes from 'prop-types';
import './metricsCompareView.scss';
import axios from 'axios';
import Multiselect from 'react-bootstrap-multiselect';
import {Well, Alert} from 'react-bootstrap';
import Select from 'react-select';
import {parseServerError} from '../Helpers/utils';
import {MetricsPlotView} from '../MetricsPlotView/metricsPlotView';
import {ProgressWrapper} from '../Helpers/hoc';

class MetricsCompareView extends Component {
  static propTypes = {
    runIds: PropTypes.arrayOf(Number).isRequired,
    isSelected: PropTypes.bool.isRequired
  };

  runIdOptionsDomNode = null;

  metricLabelOptions = [{
    label: 'Run Id',
    value: '_id'
  }, {
    label: 'Experiment Name',
    value: 'experiment.name'
  }];

  constructor(props) {
    super(props);
    this.state = {
      isLoadingRuns: false,
      metrics: [],
      runIdOptions: [],
      error: '',
      metricLabel: '_id'
    };
  }

  _initializeState = () => {
    this._loadData();
    this._populateRunIdOptions();
  };

  componentDidMount() {
    const {isSelected} = this.props;
    if (isSelected) {
      // Load data only when the component is being displayed
      this._initializeState();
    }
  }

  componentDidUpdate(prevProps, _prevState, _snapshot) {
    // Populate runId options every time runIds change
    const {runIds, isSelected} = this.props;
    if (JSON.stringify(prevProps.runIds) !== JSON.stringify(runIds) && isSelected) {
      this._initializeState();
    }
  }

  _populateRunIdOptions = () => {
    const {runIds} = this.props;
    const runIdOptions = runIds.map(runId => {
      return {
        label: runId,
        value: runId,
        selected: true
      };
    });
    this.setState({
      runIdOptions
    });
  };

  _getSelectedRunIds = runIdOptions => runIdOptions.filter(option => option.selected === true)
    .map(option => Number(option.value));

  _handleRunIdsChange = _e => {
    const runIds = this.runIdOptionsDomNode.$multiselect.val();
    this.setState(({runIdOptions}) => ({
      runIdOptions: runIdOptions.map(option => {
        option.selected = runIds.includes(String(option.value));
        return option;
      })
    }));
  };

  _loadData = () => {
    const {runIds} = this.props;
    if (runIds.length >= 2) {
      this.setState({
        isLoadingRuns: true,
        error: ''
      });
      const queryString = JSON.stringify({
        run_id: {
          $in: runIds
        }
      });
      axios.get('/api/v1/Metrics', {
        params: {
          query: queryString,
          populate: 'run'
        }
      }).then(response => {
        this.setState({
          metrics: response.data,
          isLoadingRuns: false
        });
      }).catch(error => {
        const message = parseServerError(error);
        this.setState({
          isLoadingRuns: false,
          error: message
        });
      });
    } else {
      this.setState({
        error: 'At-least two runs should be selected for comparison'
      });
    }
  };

  _handleMetricLabelChange = ({value}) => {
    this.setState({
      metricLabel: value
    });
  };

  render() {
    const {metrics, isLoadingRuns, runIdOptions, error, metricLabel} = this.state;
    const selectedRunIds = this._getSelectedRunIds(runIdOptions);
    const metricsResponseForPlot = metrics.filter(metric => selectedRunIds.includes(metric.run_id));
    const errorAlert = error ? <Alert bsStyle='danger'>{error}</Alert> : '';
    const getOption = value => {
      return this.metricLabelOptions.find(option => option.value === value);
    };

    return (
      <div className='metrics-compare-view'>
        <ProgressWrapper id='metrics-compare-progress-wrapper' loading={isLoadingRuns}>
          <div>
            {errorAlert}
            <Well bsSize='small' className='run-id-filter-well'>
              <h5>Runs: </h5>
              <div className='run-id-options-wrapper'>
                <Multiselect
                  ref={el => this.runIdOptionsDomNode = el}
                  includeSelectAllOption multiple
                  enableHTML
                  buttonText={(options, _select) => {
                    if (options.length === 0) {
                      return 'None selected';
                    }

                    return `${options.length} selected`;
                  }}
                  selectedClass='run-id-selected'
                  id='run_ids'
                  maxHeight={300}
                  data={runIdOptions}
                  onSelectAll={this._handleRunIdsChange}
                  onChange={this._handleRunIdsChange}
                  onDeselectAll={this._handleRunIdsChange}
                />
              </div>

              <h5>Metric Label: </h5>
              <div className='metric-label-options-wrapper'>
                <Select
                  className='select-metric-label'
                  test-attr='select-metric-label'
                  options={this.metricLabelOptions}
                  placeholder='Metric Label'
                  value={getOption(metricLabel)}
                  onChange={this._handleMetricLabelChange}
                />
              </div>
            </Well>
            <MetricsPlotView metricsResponse={metricsResponseForPlot} runId='' metricLabel={metricLabel}/>
          </div>
        </ProgressWrapper>
      </div>
    );
  }
}

export {MetricsCompareView};
