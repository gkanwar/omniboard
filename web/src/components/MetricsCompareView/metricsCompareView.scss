
.run-id-filter-well {
  h5 {
    display: inline-block;
    margin-left: 10px;
  }
  .run-id-options-wrapper {
    display: inline-block;
    margin-left: 10px;
  }
  .metric-label-options-wrapper {
    margin-left: 10px;
    width: 200px;
    display: inline-block;
  }
}