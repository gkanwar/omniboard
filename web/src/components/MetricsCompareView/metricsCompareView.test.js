import React from 'react';
import mockAxios from 'jest-mock-axios';
import {MetricsCompareView} from './metricsCompareView';

describe('MetricsCompareView', () => {
  let wrapper = null;
  const responseData = [
    {_id: '5a2b5a8c9c7a505a652f6127', name: 'pretrain.train.loss', run_id: 1, steps: [0, 1, 2, 3, 4],
      values: [0.7159541824544438, 0.3840367944955761, 0.3469185283233073, 0.30483262065173106, 0.28915774130337507],
      timestamps: ['2017-12-09T03:37:44.425Z', '2017-12-09T03:41:54.414Z', '2017-12-09T03:46:01.766Z', '2017-12-09T03:50:07.365Z', '2017-12-09T03:54:12.560Z']},
    {_id: '5a2b5aa09c7a505a652f6146', name: 'pretrain.val.loss', run_id: 2, steps: [0, 1, 2, 3, 4],
      values: [0.32177006650114165, 0.23237958704995795, 0.23340759051386187, 0.21925230575196739, 0.20541178824900605],
      timestamps: ['2017-12-09T03:38:01.945Z', '2017-12-09T03:42:11.673Z', '2017-12-09T03:46:18.843Z', '2017-12-09T03:50:24.377Z', '2017-12-09T03:54:29.752Z']},
    {_id: '5a2b5a8c9c7a505a652f6127', name: 'pretrain.train.loss', run_id: 3, steps: [0, 1, 2, 3, 4],
      values: [0.7859541824544438, 0.5640367944955761, 0.8769185283233073, 0.60483262065173106, 0.58915774130337507],
      timestamps: ['2017-12-09T03:37:44.425Z', '2017-12-09T03:41:54.414Z', '2017-12-09T03:46:01.766Z', '2017-12-09T03:50:07.365Z', '2017-12-09T03:54:12.560Z']},
    {_id: '5a2b5aa09c7a505a652f6146', name: 'pretrain.val.loss', run_id: 4, steps: [0, 1, 2, 3, 4],
      values: [0.52177006650114165, 0.9237958704995795, 0.11340759051386187, 0.41925230575196739, 0.90541178824900605],
      timestamps: ['2017-12-09T03:38:01.945Z', '2017-12-09T03:42:11.673Z', '2017-12-09T03:46:18.843Z', '2017-12-09T03:50:24.377Z', '2017-12-09T03:54:29.752Z']}
  ];
  const runIds = [1, 2, 3, 4];
  console.error = jest.fn();

  beforeEach(() => {
    wrapper = shallow(
      <MetricsCompareView isSelected runIds={runIds}/>
    );
  });

  afterEach(() => {
    // Cleaning up the mess left behind the previous test
    mockAxios.reset();
    jest.clearAllMocks();
    wrapper = null;
  });

  describe('should load data', () => {
    it('success', async () => {
      expect(mockAxios.get.mock.calls).toHaveLength(1);

      expect(mockAxios.get.mock.calls[0]).toEqual(['api/v1/Metrics', {params: {query: JSON.stringify({run_id: {$in: [1, 2, 3, 4]}}), populate: 'run'}}]);
      expect(wrapper.instance().state.isLoadingRuns).toBeTruthy();
      mockAxios.mockResponse({status: 200, data: responseData});

      await tick();
      expect(wrapper.update().instance().state.isLoadingRuns).toBeFalsy();
      expect(wrapper.update()).toMatchSnapshot();
    });

    it('error', () => {
      const errResponse = {status: 400, message: 'Not Found'};
      mockAxios.mockError(errResponse);

      expect(wrapper.instance().state.error).not.toEqual('');
    });
  });

  it('should not load data when runIds < 2', async () => {
    wrapper.unmount();
    mockAxios.reset();
    wrapper = mount(
      <MetricsCompareView isSelected runIds={[1]}/>
    );

    expect(mockAxios.get.mock.calls).toHaveLength(0);
    expect(wrapper.instance().state.error).not.toEqual('');
  });

  it('should handle run id change', async () => {
    mockAxios.mockResponse({status: 200, data: responseData});
    mockAxios.reset();
    wrapper.instance().runIdOptionsDomNode = {
      $multiselect: {
        val: () => ['4', '3', '1']
      }
    };
    wrapper.instance()._handleRunIdsChange();

    expect(wrapper.instance().state.runIdOptions.find(option => option.value === 2).selected).toBeFalsy();
  });

  it('should handle metric label change', async () => {
    mockAxios.mockResponse({status: 200, data: responseData});
    mockAxios.reset();
    wrapper.instance()._handleMetricLabelChange({value: 'experiment.name'});

    expect(wrapper.instance().state.metricLabel).toEqual('experiment.name');
  });

  describe('componentDidUpdate', () => {
    beforeEach(() => {
      mockAxios.mockResponse({status: 200, data: responseData});
      mockAxios.reset();
    });

    it('loads data', async () => {
      wrapper.setProps({runIds: [1, 2]});

      expect(mockAxios.get.mock.calls).toHaveLength(1);
      expect(wrapper.instance().state.isLoadingRuns).toBeTruthy();
    });
  });
});
