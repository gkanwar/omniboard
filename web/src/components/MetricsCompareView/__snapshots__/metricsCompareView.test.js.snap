// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetricsCompareView should load data success 1`] = `
<div
  className="metrics-compare-view"
>
  <ProgressWrapper
    id="metrics-compare-progress-wrapper"
    loading={false}
  >
    <div>
      <Well
        bsClass="well"
        bsSize="small"
        className="run-id-filter-well"
      >
        <h5>
          Runs: 
        </h5>
        <div
          className="run-id-options-wrapper"
        >
          <MultiSelect
            buttonText={[Function]}
            data={
              Array [
                Object {
                  "label": 1,
                  "selected": true,
                  "value": 1,
                },
                Object {
                  "label": 2,
                  "selected": true,
                  "value": 2,
                },
                Object {
                  "label": 3,
                  "selected": true,
                  "value": 3,
                },
                Object {
                  "label": 4,
                  "selected": true,
                  "value": 4,
                },
              ]
            }
            enableHTML={true}
            id="run_ids"
            includeSelectAllOption={true}
            maxHeight={300}
            multiple={true}
            onChange={[Function]}
            onDeselectAll={[Function]}
            onSelectAll={[Function]}
            selectedClass="run-id-selected"
          />
        </div>
        <h5>
          Metric Label: 
        </h5>
        <div
          className="metric-label-options-wrapper"
        >
          <StateManager
            className="select-metric-label"
            defaultInputValue=""
            defaultMenuIsOpen={false}
            defaultValue={null}
            onChange={[Function]}
            options={
              Array [
                Object {
                  "label": "Run Id",
                  "value": "_id",
                },
                Object {
                  "label": "Experiment Name",
                  "value": "experiment.name",
                },
              ]
            }
            placeholder="Metric Label"
            test-attr="select-metric-label"
            value={
              Object {
                "label": "Run Id",
                "value": "_id",
              }
            }
          />
        </div>
      </Well>
      <MetricsPlotView
        metricLabel="_id"
        metricsResponse={
          Array [
            Object {
              "_id": "5a2b5a8c9c7a505a652f6127",
              "name": "pretrain.train.loss",
              "run_id": 1,
              "steps": Array [
                0,
                1,
                2,
                3,
                4,
              ],
              "timestamps": Array [
                "2017-12-09T03:37:44.425Z",
                "2017-12-09T03:41:54.414Z",
                "2017-12-09T03:46:01.766Z",
                "2017-12-09T03:50:07.365Z",
                "2017-12-09T03:54:12.560Z",
              ],
              "values": Array [
                0.7159541824544438,
                0.3840367944955761,
                0.3469185283233073,
                0.30483262065173106,
                0.28915774130337507,
              ],
            },
            Object {
              "_id": "5a2b5aa09c7a505a652f6146",
              "name": "pretrain.val.loss",
              "run_id": 2,
              "steps": Array [
                0,
                1,
                2,
                3,
                4,
              ],
              "timestamps": Array [
                "2017-12-09T03:38:01.945Z",
                "2017-12-09T03:42:11.673Z",
                "2017-12-09T03:46:18.843Z",
                "2017-12-09T03:50:24.377Z",
                "2017-12-09T03:54:29.752Z",
              ],
              "values": Array [
                0.32177006650114165,
                0.23237958704995795,
                0.23340759051386187,
                0.21925230575196739,
                0.20541178824900605,
              ],
            },
            Object {
              "_id": "5a2b5a8c9c7a505a652f6127",
              "name": "pretrain.train.loss",
              "run_id": 3,
              "steps": Array [
                0,
                1,
                2,
                3,
                4,
              ],
              "timestamps": Array [
                "2017-12-09T03:37:44.425Z",
                "2017-12-09T03:41:54.414Z",
                "2017-12-09T03:46:01.766Z",
                "2017-12-09T03:50:07.365Z",
                "2017-12-09T03:54:12.560Z",
              ],
              "values": Array [
                0.7859541824544438,
                0.5640367944955761,
                0.8769185283233073,
                0.604832620651731,
                0.5891577413033751,
              ],
            },
            Object {
              "_id": "5a2b5aa09c7a505a652f6146",
              "name": "pretrain.val.loss",
              "run_id": 4,
              "steps": Array [
                0,
                1,
                2,
                3,
                4,
              ],
              "timestamps": Array [
                "2017-12-09T03:38:01.945Z",
                "2017-12-09T03:42:11.673Z",
                "2017-12-09T03:46:18.843Z",
                "2017-12-09T03:50:24.377Z",
                "2017-12-09T03:54:29.752Z",
              ],
              "values": Array [
                0.5217700665011417,
                0.9237958704995795,
                0.11340759051386187,
                0.41925230575196737,
                0.905411788249006,
              ],
            },
          ]
        }
        runId=""
        stateFilterKeys={
          Array [
            "selectedMetricNames",
            "selectedXAxis",
            "selectedYAxis",
            "plotWidth",
            "plotHeight",
            "smoothing",
          ]
        }
      />
    </div>
  </ProgressWrapper>
</div>
`;
